### DISCUSSION TITLE
Key arguments (_t) are not part of the State interface (v2)

### DISCUSSION THREAD

**Discussion by piotr-cz** [vercel/swr#2295] [Ideas]
# Bug report

## Description / Observed Behavior

I'm using custom cache provider to persist data in IDB.

When implementing garbage collector I'm evaluating given path (as in [example](https://github.com/piotr-cz/swr-idb-cache/tree/feature/swr-2.0#ignore-api-endpoints-from-cache-persistence)).
This works fine for cases when key argument just is simple path.

However for more complex scenarios when key arguments consists of path and query I'd like to have an option to evaluate them.

SWR v2.0 has changed [the structure of cached data](https://github.com/vercel/swr/discussions/1919) and added `_k` property which contains key arguments:

https://github.com/vercel/swr/blob/69105b4924555acff77816b57ae91d778b67fb6a/core/use-swr.ts#L97-L100

However the `_k` property is not part of the State interface:

https://github.com/vercel/swr/blob/69105b4924555acff77816b57ae91d778b67fb6a/_internal/types.ts#L288-L293

Assuming the prefix means that accessing this property is not recommended, I cannot rely on it being available in future versions of the library.

## Expected Behavior

The `_k` property is included in the State interface and it's unprefixed:

```diff
 export type State<Data = any, Error = any> = {
   data?: Data
   error?: Error
   isValidating?: boolean
   isLoading?: boolean
+  key?: Key
 }
```

## Repro Steps / Code Example

[SWR 2.0 branch for the IndexedDB Cache Provider](https://github.com/piotr-cz/swr-idb-cache/tree/feature/swr-2.0)

## Additional Context

SWR version: 2.0.0-rc.0

**Comment by koba04** [comment]
How do you use `key` in the `State`?

**Comment by piotr-cz** [comment]
The simplest case is to not store data in cache for certain paths or query params.

This is easy when key is just a simple string (`/api/user/123`): than it's possible to check on the stableHash output.

However when it's an array (`['/api/user', {userId: 123}]`) you have to access the state and it's `_k` value

**Comment by piotr-cz** [comment]
@koba04
The `key` is passed in `State` to Cache provider, although it's not defined in it's interface.

The result of executing fetcher with key `{path: 'items/news/1', query: {...}}`, in cache looks like this:

![image](https://user-images.githubusercontent.com/612953/198239833-250f4f22-cac6-4af9-8b10-2f1caf99d771.png)

For example imagine I'd want to store items in cache only when `path !== items/news/2`, or when `query.filter.organisation !== null`.

To inspect the key (path and query in example above) I can either try to decode [stableHash](https://github.com/shuding/stable-hash), or access the `State#_k` property.

- First option doesn't feel right as I assume the stableHash is used by SWR as a checksum.
- Second option involves accessing property that's not part of State interface and is prefixed.
